doc/tut3.rst

```diff
@@ -0,0 +1,277 @@
+
+.. code-block:: nim
+    :test: "nim c $1"
+
+  import macros
+
+  macro myAssert(arg: untyped): untyped =
+    # all node kind identifiers are prefixed with "nnk"
+    arg.expectKind nnkInfix
+    arg.expectLen 3
+    # operator as string literal
+    let op  = newLit(" " & arg[0].repr & " ")
+    let lhs = arg[1]
+    let rhs = arg[2]
+
+    result = quote do:
+      if not `arg`:
+        raise newException(AssertionError,$`lhs` & `op` & $`rhs`)
+
+  let a = 1
+  let b = 2
+
+  myAssert(a != b)
+  myAssert(a == b)
+
+
+This is the code that will be generated. To debug what the macro
+actually generated, the statement ``echo result.repr`` can be used, in
+the last line of the macro. It is also the statement that has been
+used to get this output.
+
+.. code-block:: nim
+  if not (a != b):
+    raise newException(AssertionError, $a & " != " & $b)
+
+With Power Comes Responsibility
+-------------------------------
+
+Macros are very powerful. A good advice is to use them as little as
+possible, but as much as necessary. Macros can change the semantics of
+expressions, making the code incomprehensible for anybody who does not
+know exactly what the macro does with it.  So whenever a macro is not
+necessary and the same logic can be implemented using templates or
+generics, it is probably better not to use a macro.  And when a macro
+is used for something, the macro should better have a well written
+documentation.  And for all the people who write only perfectly self
+explaining code, for macros the implementation is not documentation
+enough.
+
+Limitations
+-----------
+
+Since macros are evaluated in the compiler in the NimVM, macros share
+all the limitations of the NimVM. They have to be implemented in pure Nim
+code.  Macros can start external processes on the shell, but they
+cannot call C functions except from those that are built in the
+compiler.
+
+
+More Examples
+=============
+
+This tutorial can only cover the basics of the macro system. There are
+macros out there that could be an inspiration for you of what is
+possible with it.
+
+
+Strformat
+---------
+
+In the Nim standard library, the ``strformat`` library provides a
+macro that parses a string literal at compile time.  Parsing a string
+in a macro like here is generally not recommended. The parsed AST
+cannot have type information, and parsing implemented on the VM is
+generally not very fast.  Working on AST nodes is almost always the
+recommended way.  But still ``strformat`` is a good example for a
+practical use case for a macro that is slightly more complex that the
+``assert`` macro.
+
+_Strformat:https://github.com/nim-lang/Nim/blob/devel/lib/pure/strformat.nim#L280
+
+Ast Pattern Matching
+--------------------
+
+Ast Pattern Matching is a macro library to aid in writing complex
+macros.  This can be seen as a good example of how to repurpose the
+Nim syntax tree with new semantics.
+
+_Ast Pattern Matching: https://github.com/krux02/ast-pattern-matching
```

Comment:
If you're writing links like this, should you reference them beforehand somewhere in the text?

If you want to just use them at the end of the section, then the other syntax (`[backtick]MyLink <url://goes.here>[backtick]_`) is used.

Reply:
I am talking about ast pattern matching in that section. All I want is a link at the end that points to the library.

Reply:
> All I want is a link at the end that points to the library.

To get that use this syntax:
```
`Ast Pattern Matching <https://github.com/krux02/ast-pattern-matching>`_
```